File: shared/src/main/java/xyz/klinker/messenger/shared/receiver/SmsSentReceiver.java
/*
 * Copyright (C) 2017 Luke Klinker
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.klinker.messenger.shared.receiver;

import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.database.Cursor;
import android.net.Uri;
import android.provider.Telephony;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;
import android.telephony.SmsManager;
import android.util.Log;

import com.klinker.android.send_message.SentReceiver;
import com.klinker.android.send_message.StripAccents;

import java.util.HashSet;
import java.util.List;
import java.util.Set;

import xyz.klinker.messenger.shared.MessengerActivityExtras;
import xyz.klinker.messenger.shared.R;
import xyz.klinker.messenger.shared.data.ColorSet;
import xyz.klinker.messenger.shared.data.DataSource;
import xyz.klinker.messenger.shared.data.FeatureFlags;
import xyz.klinker.messenger.shared.data.Settings;
import xyz.klinker.messenger.shared.data.model.Message;
import xyz.klinker.messenger.shared.service.ResendFailedMessage;
import xyz.klinker.messenger.shared.util.ActivityUtils;
import xyz.klinker.messenger.shared.util.CursorUtil;
import xyz.klinker.messenger.shared.util.NotificationUtils;
import xyz.klinker.messenger.shared.util.SmsMmsUtils;

/**
 * Receiver for getting notifications of when an SMS has finished sending. By default it's super
 * class will mark the internal message as sent, we need to also mark our database as sent.
 */
public class SmsSentReceiver extends SentReceiver {

    protected boolean retryFailedMessages() {
        return true;
    }

    @Override
    public void onReceive(Context context, Intent intent) {
        new Thread(() -> {
            try {
                super.onReceive(context, intent);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();

        new Thread(() -> {
            try {
                handleReceiver(context, intent);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }).start();
    }

    private void handleReceiver(Context context, Intent intent) {
        Uri uri = Uri.parse(intent.getStringExtra("message_uri"));

        switch (getResultCode()) {
            case SmsManager.RESULT_ERROR_GENERIC_FAILURE:
            case SmsManager.RESULT_ERROR_NO_SERVICE:
            case SmsManager.RESULT_ERROR_NULL_PDU:
            case SmsManager.RESULT_ERROR_RADIO_OFF:
                markMessageError(context, uri);
                break;
            default:
                try {
                    markMessageSent(context, uri);
                } catch (Exception e) {
                    fallbackToLatestMessages(context);
                }
                break;
        }
    }

    private void markMessageSent(Context context, Uri uri) {
        markMessage(context, uri, false);
    }

    private void markMessageError(Context context, Uri uri) {
        markMessage(context, uri, true);
    }

    private void markMessage(Context context, Uri uri, boolean error) {
        Cursor message = SmsMmsUtils.getSmsMessage(context, uri, null);

        if (message != null && message.moveToFirst()) {
            String body = message.getString(message.getColumnIndex(Telephony.Sms.BODY));
            message.close();

            Settings settings = Settings.get(context);
            if (settings.signature != null && !settings.signature.isEmpty()) {
                body = body.replace("\n" + settings.signature, "");
            }

            DataSource source = DataSource.INSTANCE;
            Cursor messages = source.searchMessages(context, body);

            if (messages != null && messages.moveToFirst()) {
                long id = messages.getLong(0);
                long conversationId = messages
                        .getLong(messages.getColumnIndex(Message.COLUMN_CONVERSATION_ID));
                String data = messages.getString(messages.getColumnIndex(Message.COLUMN_DATA));

                markMessage(source, context, error, id, conversationId, data);
            } else {
                // if the message was unicode, then it won't match here and would never get marked as sent or error
                List<Message> messageList = source.getNumberOfMessages(context, 10);
                boolean markedAsSent = false;
                for (Message m : messageList) {
                    if (StripAccents.stripAccents(m.data).equals(body) && m.type == Message.TYPE_SENDING) {
                        markMessage(source, context, error, m.id, m.conversationId, m.data);
                        markedAsSent = true;
                        break;
                    }
                }

                if (!markedAsSent) {
                    Set<Long> conversationIds = new HashSet<>();

                    for (Message m : messageList) {
                        if (m.type == Message.TYPE_SENDING) {
                            source.updateMessageType(context, m.id, error ? Message.TYPE_ERROR : Message.TYPE_SENT);
                            conversationIds.add(m.conversationId);
                        }
                    }

                    for (Long id : conversationIds) {
                        MessageListUpdatedReceiver.sendBroadcast(context, id);
                    }
                }
            }

            CursorUtil.closeSilent(messages);
        } else {
            throw new RuntimeException("no messages found");
        }
    }

    private void markMessage(DataSource source, Context context, boolean error, long messageId, long conversationId, String data) {
        source.updateMessageType(context, messageId, error ? Message.TYPE_ERROR : Message.TYPE_SENT);

        MessageListUpdatedReceiver.sendBroadcast(context, conversationId);

        Intent resend = new Intent(context, ResendFailedMessage.class);
        resend.putExtra(ResendFailedMessage.EXTRA_MESSAGE_ID, messageId);
        resend.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);

        if (error) {
            if (FeatureFlags.get(context).AUTO_RETRY_FAILED_MESSAGES && retryFailedMessages()) {
                context.startService(resend);
            } else {
                Intent open = ActivityUtils.buildForComponent(ActivityUtils.MESSENGER_ACTIVITY);
                open.putExtra(MessengerActivityExtras.INSTANCE.getEXTRA_CONVERSATION_ID(), conversationId);
                open.putExtra(MessengerActivityExtras.INSTANCE.getEXTRA_FROM_NOTIFICATION(), true);
                open.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
                PendingIntent pendingOpen = PendingIntent.getActivity(context,
                        (int) conversationId, open, PendingIntent.FLAG_UPDATE_CURRENT);

                NotificationCompat.Builder notification = new NotificationCompat.Builder(context, NotificationUtils.TEST_NOTIFICATIONS_CHANNEL_ID)
                        .setSmallIcon(R.drawable.ic_stat_notify)
                        .setContentTitle(context.getString(R.string.message_sending_failed))
                        .setContentText(data)
                        .setColor(ColorSet.DEFAULT(context).color)
                        .setAutoCancel(true)
                        .setContentIntent(pendingOpen);

                PendingIntent resendPending = PendingIntent.getService(context,
                        (int) messageId, resend, PendingIntent.FLAG_UPDATE_CURRENT);
                NotificationCompat.Action action = new NotificationCompat.Action(
                        R.drawable.ic_reply_dark, context.getString(R.string.resend), resendPending);

                notification.addAction(action);
                NotificationManagerCompat.from(context)
                        .notify(6666 + (int) messageId, notification.build());
            }
        }
    }

    public void fallbackToLatestMessages(Context context) {
        try {
            DataSource source = DataSource.INSTANCE;
            List<Message> messageList = source.getNumberOfMessages(context, 10);

            Set<Long> conversationIds = new HashSet<>();

            for (Message m : messageList) {
                if (m.type == Message.TYPE_SENDING) {
                    source.updateMessageType(context, m.id, Message.TYPE_SENT);
                    conversationIds.add(m.conversationId);
                }
            }

            for (Long id : conversationIds) {
                MessageListUpdatedReceiver.sendBroadcast(context, id);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: shared/src/main/java/xyz/klinker/messenger/shared/receiver/SmsSentReceiver.kt
/*
 * Copyright (C) 2020 Luke Klinker
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package xyz.klinker.messenger.shared.receiver

import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.net.Uri
import android.provider.Telephony
import androidx.core.app.NotificationCompat
import androidx.core.app.NotificationManagerCompat
import android.telephony.SmsManager
import com.klinker.android.send_message.SentReceiver
import com.klinker.android.send_message.StripAccents
import xyz.klinker.messenger.api.implementation.Account
import xyz.klinker.messenger.shared.MessengerActivityExtras
import xyz.klinker.messenger.shared.R
import xyz.klinker.messenger.shared.data.ColorSet
import xyz.klinker.messenger.shared.data.DataSource
import xyz.klinker.messenger.shared.data.Settings
import xyz.klinker.messenger.shared.data.model.Message
import xyz.klinker.messenger.shared.service.ResendFailedMessage
import xyz.klinker.messenger.shared.util.*
import java.util.*

/**
 * Receiver for getting notifications of when an SMS has finished sending. By default it's super
 * class will mark the internal message as sent, we need to also mark our database as sent.
 */
open class SmsSentReceiver : SentReceiver() {

    protected open fun retryFailedMessages() = false

    override fun updateInInternalDatabase(context: Context, intent: Intent, resultCode: Int) {
        Thread { super.updateInInternalDatabase(context, intent, resultCode) }.start()
    }

    override fun onMessageStatusUpdated(context: Context, intent: Intent, receiverResultCode: Int) {
        if (Account.exists() && !Account.primary) {
            return
        }

        try {
            val uri = Uri.parse(intent.getStringExtra("message_uri"))
            when (receiverResultCode) {
                SmsManager.RESULT_ERROR_GENERIC_FAILURE, SmsManager.RESULT_ERROR_NO_SERVICE, SmsManager.RESULT_ERROR_NULL_PDU, SmsManager.RESULT_ERROR_RADIO_OFF -> markMessageError(context, uri)
                else -> try {
                    markMessageSent(context, uri)
                } catch (e: Exception) {
                    e.printStackTrace()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    private fun markMessageSent(context: Context, uri: Uri) {
//        markMessage(context, uri, false)
        markLatestAsRead(context)
    }

    private fun markMessageError(context: Context, uri: Uri) {
        markMessage(context, uri, true)
    }

    private fun markMessage(context: Context, uri: Uri, error: Boolean) {
        val message = SmsMmsUtils.getSmsMessage(context, uri, null)

        if (message != null && message.moveToFirst()) {
            var body = message.getString(message.getColumnIndex(Telephony.Sms.BODY))
            message.close()

            if (Settings.signature != null && !Settings.signature!!.isEmpty()) {
                body = body.replace("\n" + Settings.signature!!, "")
            }

            val source = DataSource
            val messages = source.searchMessages(context, body)

            if (messages != null && messages.moveToFirst()) {
                val id = messages.getLong(0)
                val conversationId = messages
                        .getLong(messages.getColumnIndex(Message.COLUMN_CONVERSATION_ID))
                val data = messages.getString(messages.getColumnIndex(Message.COLUMN_DATA))

                markMessage(source, context, error, id, conversationId, data)
            } else {
                // if the message was unicode, then it won't match here and would never get marked as sent or error
                val messageList = source.getNumberOfMessages(context, 10)
                var markedAsSent = false
                for (m in messageList) {
                    if (StripAccents.stripAccents(m.data) == body && m.type == Message.TYPE_SENDING) {
                        markMessage(source, context, error, m.id, m.conversationId, m.data)
                        markedAsSent = true
                        break
                    }
                }

                if (!markedAsSent) {
                    val conversationIds = HashSet<Long>()

                    for (m in messageList) {
                        if (m.type == Message.TYPE_SENDING) {
                            source.updateMessageType(context, m.id, if (error) Message.TYPE_ERROR else Message.TYPE_SENT)
                            conversationIds.add(m.conversationId)
                        }
                    }

                    for (id in conversationIds) {
                        MessageListUpdatedReceiver.sendBroadcast(context, id)
                    }
                }
            }

            messages?.closeSilent()
        } else {
            throw RuntimeException("no messages found")
        }
    }

    private fun markMessage(source: DataSource, context: Context, error: Boolean, messageId: Long, conversationId: Long, data: String?) {
        source.updateMessageType(context, messageId, if (error) Message.TYPE_ERROR else Message.TYPE_SENT)

        MessageListUpdatedReceiver.sendBroadcast(context, conversationId)

        val resend = Intent(context, ResendFailedMessage::class.java)
        resend.putExtra(ResendFailedMessage.EXTRA_MESSAGE_ID, messageId)
        resend.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK

        if (error) {
            val conversation = DataSource.getConversation(context, conversationId)
            if (conversation != null) {
                conversation.snippet = conversation.snippet!!.replace("You: ", "")
                conversation.snippet = context.getString(R.string.failed) + ": " + conversation.snippet

                DataSource.updateConversationSnippet(context, conversationId, conversation.timestamp, conversation.snippet!!, true)
                ConversationListUpdatedReceiver.sendBroadcast(context, conversationId, conversation.snippet, true)
            }

            if (retryFailedMessages()) {
                context.startService(resend)
            } else {
                val open = ActivityUtils.buildForComponent(ActivityUtils.MESSENGER_ACTIVITY)
                open.putExtra(MessengerActivityExtras.EXTRA_CONVERSATION_ID, conversationId)
                open.putExtra(MessengerActivityExtras.EXTRA_FROM_NOTIFICATION, true)
                open.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
                val pendingOpen = PendingIntent.getActivity(context,
                        conversationId.toInt(), open, PendingIntent.FLAG_UPDATE_CURRENT)

                val notification = NotificationCompat.Builder(context, NotificationUtils.DEFAULT_CONVERSATION_CHANNEL_ID)
                        .setSmallIcon(R.drawable.ic_stat_notify)
                        .setContentTitle(context.getString(R.string.message_sending_failed))
                        .setContentText(data)
                        .setColor(ColorSet.DEFAULT(context).color)
                        .setAutoCancel(true)
                        .setContentIntent(pendingOpen)

                val resendPending = PendingIntent.getService(context,
                        messageId.toInt(), resend, PendingIntent.FLAG_UPDATE_CURRENT)
                val action = NotificationCompat.Action(
                        R.drawable.ic_reply_dark, context.getString(R.string.resend), resendPending)

                notification.addAction(action)
                NotificationManagerCompat.from(context)
                        .notify(6666 + messageId.toInt(), notification.build())
            }
        }
    }

    companion object {
        fun markLatestAsRead(context: Context) {
            // if we get dropped into this, we are just going to mark the latest messages as sent, instead of
            // trying to query them from the shared database and mark them as sent like we previously did.
            // this is a hack, since we are not storing the id of the messages in the shared database.

            val messageList = DataSource.getNumberOfMessages(context, 10)
            val conversationIds = mutableSetOf<Long>()

            messageList.forEach {
                if (it.type == Message.TYPE_SENDING) {
                    DataSource.updateMessageType(context, it.id, Message.TYPE_SENT)
                    conversationIds.add(it.conversationId)
                }
            }

            conversationIds.forEach {
                MessageListUpdatedReceiver.sendBroadcast(context, it)
            }
        }
    }
}
